text_scrubber/geo/geo.py

```diff
@@ -0,0 +1,336 @@
+    resources['cleaned_trigrams'] = {cleaned_country: get_trigram_tokens(cleaned_country)
+                                     for cleaned_country in resources['cleaned_to_capitilized'].keys()}
+
+    return resources
+
+
+_COUNTRY_RESOURCES = _get_country_resources()
+
+
+def _get_state_resources() -> Dict[str, Dict]:
+    """
+    Reads and parses state resource files.
+
+    :return: Dictionary containing resources used for state normalization.
+    """
+    resources = dict()
+
+    # Read in a map of states per country. For some states the state code is added. States and/or state codes are not
+    # always unique. We don't remove stop words when we're dealing with state codes.
+    resources['state_country_map'] = defaultdict(set)
+    state_country_file = (line.split(", ") for line in read_resource_file(__file__, 'resources/states_per_country.txt'))
+    for *state_list, canonical_country in state_country_file:
+        for _state in state_list:
+            resources['state_country_map'][clean_state(_state)].add((capitalize_geo_string(state_list[0]),
+                                                                     clean_country(canonical_country)))
+
+    # Generate trigrams for the cleaned states
+    resources['cleaned_trigrams'] = {cleaned_state: get_trigram_tokens(cleaned_state)
+                                     for cleaned_state in resources['state_country_map'].keys()}
+
+    return resources
+
+
+_STATE_RESOURCES = _get_state_resources()
+
+
+# TODO: this is the old version
+# def _get_city_resources() -> Dict[str, Dict]:
+#     """
+#     Reads and parses city resource files.
+#
+#     :return: Dictionary containing resources used for city normalization.
+#     """
+#     resources = dict()
+#
+#     # Get a map of cleaned city name to (capitalized city name, cleaned country). City names are not always unique. Some
+#     # countries here first have to be normalized
+#     resources['city_country_map'] = defaultdict(set)
+#     city_country_file = (line.split(", ") for line in read_resource_file(__file__, 'resources/cities_per_country.txt'))
+#     for canonical_country, city in city_country_file:
+#         cleaned_country = clean_country(canonical_country)
+#         if cleaned_country not in _COUNTRY_RESOURCES['cleaned_to_capitilized']:
+#             cleaned_country = [clean_country(country) for country in normalize_country(canonical_country)]
+#         else:
+#             cleaned_country = [cleaned_country]
+#         for country in cleaned_country:
+#             resources['city_country_map'][clean_city(city)].add((capitalize_geo_string(city), country))
+#
+#     # Generate trigrams for the cleaned states
+#     resources['cleaned_trigrams'] = {cleaned_city: get_trigram_tokens(cleaned_city)
+#                                      for cleaned_city in resources['city_country_map'].keys()}
+#
+#     return resources
+
+
+def _get_city_resources(restrict_countries_code: Optional[Set] = None) -> Dict[str, Dict]:
```

Comment:
Let's add an additional function here called `add_city_resources` which takes `countries` as input and can be called with any number of countries (be it country or country code). It will then convert the input to a list of country codes and calls `_get_city_resources` to retrieve the appropriate resources for the countries it hasn't loaded before and subsequently updates `_CITY_RESOURCES`.

Then that function can be called by `normalize_city` whenever the `restrict_countries` is passed.